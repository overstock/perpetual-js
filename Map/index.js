import Map, { MapPrototype } from './Map';
import {
  getIn,
  hashCode,
  hasIn,
  map,
  merge,
  mergeDeep,
  mergeDeepIn,
  mergeIn,
  remove,
  removeIn,
  setIn,
  spread,
  toJS,
  toMethod,
  update,
  updateIn,
  withMutations,
} from '../_methods';

MapPrototype.getIn = toMethod(getIn);
MapPrototype.hashCode = toMethod(hashCode);
MapPrototype.hasIn = toMethod(hasIn);
MapPrototype.map = toMethod(map);
MapPrototype.merge = toMethod(merge);
MapPrototype.mergeDeep = toMethod(mergeDeep);
MapPrototype.mergeDeepIn = toMethod(mergeDeepIn);
MapPrototype.mergeIn = toMethod(mergeIn);
MapPrototype.delete = toMethod(remove);
MapPrototype.deleteIn = toMethod(removeIn);
MapPrototype.setIn = toMethod(setIn);
MapPrototype.spread = toMethod(spread);
MapPrototype.toJS = toMethod(toJS);
MapPrototype.update = toMethod(update);
MapPrototype.updateIn = toMethod(updateIn);
MapPrototype.withMutations = withMutations;

const Mep = value => new Map(value);

export default Mep;
