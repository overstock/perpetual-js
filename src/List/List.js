import { isList } from '../Perpetual';
import { IS_LIST_SYMBOL } from '../Perpetual/isList';
import {
  SIZE, makeRef, setRef, NOT_SET,
} from '../_methods/utils/TrieUtils';
import { freeze } from '../_methods/utils';

class List {
  constructor(value) {
    const empty = emptyList();
    if (value === null || value === undefined) return empty;
    if (isList(value)) return value;
    if (!Array.isArray(value)) {
      throw new TypeError(`Expected Array: ${value}`);
    }
    const { length } = value;
    if (length === 0) return empty;
    if (length > SIZE) throw new SyntaxError(`Array.length needs to be less than: ${SIZE}`);
    if (length > 0 && length < SIZE) {
      return makeList(length, new VNode(value));
    }
  }

  has(index) {
    if (typeof index !== 'number') return;
    if (index < 0 || index > this.size) return false;
    return true;
  }

  get(index, notSetValue) {
    if (typeof index !== 'number') return;
    const { size, tail } = this;
    if (index >= 0 && index < size) {
      return tail.get(index, notSetValue);
    }
    return notSetValue;
  }

  set(index, value) {
    return updateList(this, index, value);
  }

  reduce(fn, t) {
    return this.tail ? this.tail.reduce(fn, t) : NOT_SET;
  }
}

List.isList = isList;

const ListPrototype = List.prototype;
ListPrototype[IS_LIST_SYMBOL] = true;

class VNode {
  constructor(array) {
    this.array = freeze(array);
  }

  get(index, notSetValue) {
    const { array } = this;
    if (index >= 0 && index < array.length) {
      return array[index];
    }
    return notSetValue;
  }

  reduce(fn, t) {
    const { array } = this;
    return Array.isArray(array) ? array.reduce(fn, t) : NOT_SET;
  }
}

let EMPTY_LIST;
const emptyList = () => {
  if (!EMPTY_LIST) EMPTY_LIST = makeList(0);
  return EMPTY_LIST;
};

const makeList = (size, tail, hashCode) => {
  const list = Object.create(ListPrototype);
  list.size = size;
  list.tail = tail;
  list.altered = false;
  list._hashCode = hashCode;
  return list;
};

const updateList = (list, index, value) => {
  if (index !== index) return list; // eslint-disable-line
  if (index < 0) throw new SyntaxError('Index must be greater than or equal to 0');
  const { size } = list;
  const didAlter = makeRef();
  const newTail = createNewTail(list, index, value, didAlter);
  const newSize = newTail ? newTail.length : size;
  if (!didAlter.value) return list;
  return makeList(newSize, new VNode(newTail));
};

const createNewTail = (list, index, value, didAlter) => {
  let { tail } = list;
  if (!tail) tail = new VNode([]);
  const current = list.get(index);
  if (current === value) {
    return;
  }
  setRef(didAlter);
  return Object.assign([...tail.array], { [index]: value });
};

export {
  List as default,
  ListPrototype,
  VNode,
  emptyList,
  makeList,
  updateList,
  createNewTail,
};
