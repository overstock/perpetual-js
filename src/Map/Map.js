import { isMap } from '../Perpetual';
import { NOT_SET, setRef, makeRef } from '../_methods/utils/TrieUtils';
import { IS_MAP_SYMBOL } from '../Perpetual/isMap';
import {
  arrCopy,
  freeze,
  is,
  isPlainObject,
  objectToArray,
} from '../_methods/utils';

class Map {
  constructor(value) {
    if (value === null || value === undefined) return emptyMap();
    if (isMap(value)) return value;
    return createNewMap(value);
  }

  get(key, notSetValue) {
    return this.root ? this.root.get(key, notSetValue) : notSetValue;
  }

  has(searchKey) {
    return this.get(searchKey, NOT_SET) !== NOT_SET;
  }

  set(key, value) {
    return updateMap(this, key, value);
  }

  remove(key) {
    return updateMap(this, key, NOT_SET);
  }

  reduce(fn, t) {
    return this.root ? this.root.reduce(fn, t) : NOT_SET;
  }
}

Map.isMap = isMap;

const MapPrototype = Map.prototype;
MapPrototype[IS_MAP_SYMBOL] = true;

class ArrayMapNode {
  constructor(entries) {
    this.entries = freeze(entries);
  }

  get(key, notSetValue) {
    const { entries } = this;
    for (let i = 0; i < entries.length; i++) {
      if (is(key, entries[i][0])) return entries[i][1];
    }
    return notSetValue;
  }

  update(key, value, didChangeSize) {
    const { entries } = this;
    const { length } = entries;
    const removed = value === NOT_SET;
    let idx = 0;

    for (; idx < length; idx++) {
      if (is(key, entries[idx][0])) break;
    }
    const exists = idx < length;

    if (exists ? entries[idx][1] === value : removed) return this;
    if (removed || !exists) setRef(didChangeSize);
    if (removed && entries.length === 1) return; // undefined

    const newEntries = arrCopy(entries);

    if (exists) {
      if (removed) {
        if (idx === length - 1) newEntries.pop();
        else newEntries[idx] = newEntries.pop();
      } else newEntries[idx] = [key, value];
    } else newEntries.push([key, value]);

    return new ArrayMapNode(newEntries);
  }

  reduce(fn, t) {
    const { entries } = this;
    return Array.isArray(entries) ? entries.reduce(fn, t) : NOT_SET;
  }
}

let EMPTY_MAP;
const emptyMap = () => {
  if (!EMPTY_MAP) EMPTY_MAP = makeMap(0);
  return EMPTY_MAP;
};

const makeMap = (size, root) => {
  const map = Object.create(MapPrototype);
  map.size = size;
  map.root = root;
  map.altered = false;
  return freeze(map);
};

const createNewMap = value => {
  if (!isPlainObject(value)) {
    throw new Error('expected value of type object or array');
  }

  const entries = objectToArray(value);
  const map = Object.create(MapPrototype);
  map.size = entries.length;
  map.root = new ArrayMapNode(entries);
  map.altered = false;
  return map;
};

const updateMap = (map, key, value) => {
  let newRoot;
  let newSize;
  if (!map.root) {
    if (value === NOT_SET) return map;
    newSize = 1;
    newRoot = new ArrayMapNode([[key, value]]);
  } else {
    const didChangeSize = makeRef();
    newRoot = map.root.update(key, value, didChangeSize);
    let changedBy;
    if (didChangeSize.value) {
      if (value === NOT_SET) changedBy = -1;
      else changedBy = 1;
    } else changedBy = 0;
    newSize = map.size + changedBy;
  }
  return newRoot ? makeMap(newSize, newRoot) : emptyMap();
};

export {
  Map as default,
  MapPrototype,
  ArrayMapNode,
  emptyMap,
  makeMap,
  createNewMap,
  updateMap,
};
