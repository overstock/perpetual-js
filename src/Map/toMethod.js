function toMethod(func, ...args) {
  return func(this, ...args);
}

export default toMethod;
