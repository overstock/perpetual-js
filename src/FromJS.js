import Map from './Map';
import List from './List';
import { isDataStructure, isPlainObject } from './_methods/utils';

const fromJS = collection => {
  if (!isDataStructure) return collection;
  if (isPlainObject(collection)) return mapIt(collection);
  if (Array.isArray(collection)) return listIt(collection);
  return collection;
};

const mapIt = collection => {
  const map = Map(collection);
  return map.reduce((total, [key, value]) => {
    if (isPlainObject(value)) value = mapIt(value);
    else if (Array.isArray(value)) value = listIt(value);
    const result = total.set(key, value);
    return result;
  }, map);
};

const listIt = collection => {
  const list = List(collection);
  return list.reduce((total, entry, index) => {
    if (Array.isArray(entry)) entry = listIt(entry);
    else if (isPlainObject(entry)) entry = mapIt(entry);
    const result = total.set(index, entry);
    return result;
  }, list);
};

export default fromJS;
