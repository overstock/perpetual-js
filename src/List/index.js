import List, { ListPrototype } from './List';
import {
  clear,
  deLete,
  getIn,
  includes,
  insert,
  merge,
  mergeDeepIn,
  mergeIn,
  pop,
  push,
  setIn,
  shift,
  splice,
  toJS,
  toMethod,
  unshift,
  update,
  updateIn,
  withMutations,
} from '../_methods';

ListPrototype.clear = clear;
ListPrototype.concat = toMethod(merge);
ListPrototype.delete = toMethod(deLete);
ListPrototype.getIn = toMethod(getIn);
ListPrototype.includes = toMethod(includes);
ListPrototype.insert = toMethod(insert);
ListPrototype.merge = toMethod(merge);
ListPrototype.mergeDeepIn = toMethod(mergeDeepIn);
ListPrototype.mergeIn = toMethod(mergeIn);
ListPrototype.pop = toMethod(pop);
ListPrototype.push = toMethod(push);
ListPrototype.setIn = toMethod(setIn);
ListPrototype.shift = toMethod(shift);
ListPrototype.splice = toMethod(splice);
ListPrototype.toJS = toMethod(toJS);
ListPrototype.unshift = toMethod(unshift);
ListPrototype.update = toMethod(update);
ListPrototype.updateIn = toMethod(updateIn);
ListPrototype.withMutations = toMethod(withMutations);

const Lest = value => new List(value);

export {
  Lest as default,
};
