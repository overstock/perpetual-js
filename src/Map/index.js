import toMethod from './toMethod';
import { IS_MAP_SYMBOL } from '../_methods/isMap';
import {
  isMap,
  getIn,
  remove,
  removeIn,
  setIn,
  update,
  updateIn,
  withMutations,
  merge,
  mergeDeep,
  mergeIn,
} from '../_methods';
import {
  freeze, isObject, objectToArray,
} from './functions/functions';
import { is } from '../Perpetual';
import { arrCopy } from '../_methods/utils';
import { NOT_SET, makeRef, setRef } from '../_methods/utils/TrieUtils';

class Map {
  constructor(value) {
    return value === null || value === undefined
      ? emptyMap()
      : isMap(value)
        ? value
        : createNewMap(value);
  }

  get(key, notSetValue) {
    return this.root ? this.root.get(key, notSetValue) : notSetValue;
  }

  has(searchKey) {
    return this.get(searchKey, NOT_SET) !== NOT_SET;
  }

  set(key, value) {
    return updateMap(this, key, value);
  }

  remove(key) {
    return updateMap(this, key, NOT_SET);
  }

  reduce(fn, t) {
    return this.root ? this.root.reduce(fn, t) : NOT_SET;
  }
}

Map.isMap = isMap;

const MapPrototype = Map.prototype;
MapPrototype[IS_MAP_SYMBOL] = true;
MapPrototype.getIn = toMethod(getIn);
MapPrototype.delete = toMethod(remove);
MapPrototype.deleteIn = toMethod(removeIn);
MapPrototype.setIn = toMethod(setIn);
MapPrototype.update = toMethod(update);
MapPrototype.updateIn = toMethod(updateIn);
MapPrototype.merge = toMethod(merge);
MapPrototype.mergeDeep = toMethod(mergeDeep);
MapPrototype.mergeIn = toMethod(mergeIn);
MapPrototype.withMutations = withMutations;

class ArrayMapNode {
  constructor(entries) {
    this.entries = freeze(entries);
  }

  get(key, notSetValue) {
    const { entries } = this;
    for (let i = 0; i < entries.length; i++) {
      if (is(key, entries[i][0])) return entries[i][1];
    }
    return notSetValue;
  }

  update(key, value, didChangeSize) {
    const { entries } = this;
    const { length } = entries;
    const removed = value === NOT_SET;
    let idx = 0;

    for (; idx < length; idx++) {
      if (is(key, entries[idx][0])) break;
    }
    const exists = idx < length;

    if (exists ? entries[idx][1] === value : removed) return this;
    if (removed || !exists) setRef(didChangeSize);
    if (removed && entries.length === 1) return; // undefined

    const newEntries = arrCopy(entries);

    if (exists) {
      if (removed) {
        if (idx === length - 1) newEntries.pop();
        else newEntries[idx] = newEntries.pop();
      } else newEntries[idx] = [key, value];
    } else newEntries.push([key, value]);

    return new ArrayMapNode(newEntries);
  }

  reduce(fn, t) {
    const { entries } = this;
    return Array.isArray(entries) ? entries.reduce(fn, t) : NOT_SET;
  }
}

let EMPTY_MAP;
const emptyMap = () => {
  if (!EMPTY_MAP) EMPTY_MAP = makeMap(0);
  return EMPTY_MAP;
};

const makeMap = (size, root) => {
  const map = Object.create(MapPrototype);
  map.size = size;
  map.root = root;
  map.altered = false;
  return freeze(map);
};

const createNewMap = value => {
  if (!isObject(value)) {
    throw new Error('expected value of type object or array');
  }

  const entries = objectToArray(value);
  const map = Object.create(MapPrototype);
  map.size = entries.length;
  map.root = new ArrayMapNode(entries);
  map.altered = false;
  return map;
};

const updateMap = (map, key, value) => {
  let newRoot;
  let newSize;
  if (!map.root) {
    if (value === NOT_SET) return map;
    newSize = 1;
    newRoot = new ArrayMapNode([[key, value]]);
  } else {
    const didChangeSize = makeRef();
    newRoot = map.root.update(key, value, didChangeSize);
    newSize = map.size + (didChangeSize.value ? (value === NOT_SET ? -1 : 1) : 0);
  }
  return newRoot ? makeMap(newSize, newRoot) : emptyMap();
};

const Mep = value => new Map(value);

export {
  Mep as default, emptyMap, makeMap, createNewMap, ArrayMapNode,
};
