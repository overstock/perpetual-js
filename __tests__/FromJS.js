import FromJS from '../src/FromJS';
import Map from '../src/Map';
import List from '../src/List';

describe('FromJS', () => {
  describe('object to map', () => {
    const obj = {
      x: 'xyz',
      y: {
        z: 'value',
        x: 'string'
      },
      z: {
        a: {
          b: 'bc',
          c: {
            x: 'y',
            z: 'xyx',
            y: {
              a: 'abc',
              c: {
                d: 'dad',
                m: 'mom',
                g: {
                  x: 'new'
                }
              }
            }
          }
        }
      }
    };
    const fromJSObj = FromJS(obj);

    test('returns map with all values', () => {
      expect(fromJSObj).toMatchSnapshot();
    });
  });

  describe('array to List', () => {
    const arr = [
      1, 2, 3, [
        1, 2, 3, [
          1, 2, 3, [
            1, 2, 3, [
              1, 2, 3, [
                4, 5, 6, [
                  7, 8, 9, [
                    0
                  ]
                ]
              ]
            ]
          ]
        ]
      ]
    ];
    const fromJSArr = FromJS(arr);

    test('retuns list with all values', () => {
      expect(fromJSArr).toMatchSnapshot();
    });
  });

  describe('array and object to Perpetual', () => {
    const data = [
      1, 2, 3, {
        x: 'xyz',
        y: [
          1, 2, 3, [
            1, 2, 3
          ]
        ],
        z: {
          a: 'abc',
          b: {
            c: [
              1, 2, 3, {
                d: [
                  4, 5, 6, {
                    e: [
                      7, 8, 9, [
                        0
                      ]
                    ]
                  }
                ]
              }
            ]
          }
        }
      },
      [4, 5, 6]
    ];
    const fromJSData = FromJS(data);

    test('collection of Map and List containing correct values', () => {
      expect(fromJSData).toMatchSnapshot();
    });
  });
});
