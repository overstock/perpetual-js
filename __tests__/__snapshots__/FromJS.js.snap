// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FromJS array and object to Perpetual collection of Map and List containing correct values 1`] = `
List {
  "_hashCode": undefined,
  "altered": false,
  "size": 5,
  "tail": VNode {
    "array": Array [
      1,
      2,
      3,
      Map {
        "_hashCode": undefined,
        "altered": false,
        "root": ArrayMapNode {
          "entries": Array [
            Array [
              "x",
              "xyz",
            ],
            Array [
              "y",
              List {
                "_hashCode": undefined,
                "altered": false,
                "size": 4,
                "tail": VNode {
                  "array": Array [
                    1,
                    2,
                    3,
                    List {
                      "_hashCode": undefined,
                      "altered": false,
                      "size": 3,
                      "tail": VNode {
                        "array": Array [
                          1,
                          2,
                          3,
                        ],
                      },
                    },
                  ],
                },
              },
            ],
            Array [
              "z",
              Map {
                "_hashCode": undefined,
                "altered": false,
                "root": ArrayMapNode {
                  "entries": Array [
                    Array [
                      "a",
                      "abc",
                    ],
                    Array [
                      "b",
                      Map {
                        "_hashCode": undefined,
                        "altered": false,
                        "root": ArrayMapNode {
                          "entries": Array [
                            Array [
                              "c",
                              List {
                                "_hashCode": undefined,
                                "altered": false,
                                "size": 4,
                                "tail": VNode {
                                  "array": Array [
                                    1,
                                    2,
                                    3,
                                    Map {
                                      "_hashCode": undefined,
                                      "altered": false,
                                      "root": ArrayMapNode {
                                        "entries": Array [
                                          Array [
                                            "d",
                                            List {
                                              "_hashCode": undefined,
                                              "altered": false,
                                              "size": 4,
                                              "tail": VNode {
                                                "array": Array [
                                                  4,
                                                  5,
                                                  6,
                                                  Map {
                                                    "_hashCode": undefined,
                                                    "altered": false,
                                                    "root": ArrayMapNode {
                                                      "entries": Array [
                                                        Array [
                                                          "e",
                                                          List {
                                                            "_hashCode": undefined,
                                                            "altered": false,
                                                            "size": 4,
                                                            "tail": VNode {
                                                              "array": Array [
                                                                7,
                                                                8,
                                                                9,
                                                                List {
                                                                  "_hashCode": undefined,
                                                                  "altered": false,
                                                                  "size": 1,
                                                                  "tail": VNode {
                                                                    "array": Array [
                                                                      0,
                                                                    ],
                                                                  },
                                                                },
                                                              ],
                                                            },
                                                          },
                                                        ],
                                                      ],
                                                    },
                                                    "size": 1,
                                                  },
                                                ],
                                              },
                                            },
                                          ],
                                        ],
                                      },
                                      "size": 1,
                                    },
                                  ],
                                },
                              },
                            ],
                          ],
                        },
                        "size": 1,
                      },
                    ],
                  ],
                },
                "size": 2,
              },
            ],
          ],
        },
        "size": 3,
      },
      List {
        "_hashCode": undefined,
        "altered": false,
        "size": 3,
        "tail": VNode {
          "array": Array [
            4,
            5,
            6,
          ],
        },
      },
    ],
  },
}
`;

exports[`FromJS array to List retuns list with all values 1`] = `
List {
  "_hashCode": undefined,
  "altered": false,
  "size": 4,
  "tail": VNode {
    "array": Array [
      1,
      2,
      3,
      List {
        "_hashCode": undefined,
        "altered": false,
        "size": 4,
        "tail": VNode {
          "array": Array [
            1,
            2,
            3,
            List {
              "_hashCode": undefined,
              "altered": false,
              "size": 4,
              "tail": VNode {
                "array": Array [
                  1,
                  2,
                  3,
                  List {
                    "_hashCode": undefined,
                    "altered": false,
                    "size": 4,
                    "tail": VNode {
                      "array": Array [
                        1,
                        2,
                        3,
                        List {
                          "_hashCode": undefined,
                          "altered": false,
                          "size": 4,
                          "tail": VNode {
                            "array": Array [
                              1,
                              2,
                              3,
                              List {
                                "_hashCode": undefined,
                                "altered": false,
                                "size": 4,
                                "tail": VNode {
                                  "array": Array [
                                    4,
                                    5,
                                    6,
                                    List {
                                      "_hashCode": undefined,
                                      "altered": false,
                                      "size": 4,
                                      "tail": VNode {
                                        "array": Array [
                                          7,
                                          8,
                                          9,
                                          List {
                                            "_hashCode": undefined,
                                            "altered": false,
                                            "size": 1,
                                            "tail": VNode {
                                              "array": Array [
                                                0,
                                              ],
                                            },
                                          },
                                        ],
                                      },
                                    },
                                  ],
                                },
                              },
                            ],
                          },
                        },
                      ],
                    },
                  },
                ],
              },
            },
          ],
        },
      },
    ],
  },
}
`;

exports[`FromJS object to map returns map with all values 1`] = `
Map {
  "_hashCode": undefined,
  "altered": false,
  "root": ArrayMapNode {
    "entries": Array [
      Array [
        "x",
        "xyz",
      ],
      Array [
        "y",
        Map {
          "_hashCode": undefined,
          "altered": false,
          "root": ArrayMapNode {
            "entries": Array [
              Array [
                "z",
                "value",
              ],
              Array [
                "x",
                "string",
              ],
            ],
          },
          "size": 2,
        },
      ],
      Array [
        "z",
        Map {
          "_hashCode": undefined,
          "altered": false,
          "root": ArrayMapNode {
            "entries": Array [
              Array [
                "a",
                Map {
                  "_hashCode": undefined,
                  "altered": false,
                  "root": ArrayMapNode {
                    "entries": Array [
                      Array [
                        "b",
                        "bc",
                      ],
                      Array [
                        "c",
                        Map {
                          "_hashCode": undefined,
                          "altered": false,
                          "root": ArrayMapNode {
                            "entries": Array [
                              Array [
                                "x",
                                "y",
                              ],
                              Array [
                                "z",
                                "xyx",
                              ],
                              Array [
                                "y",
                                Map {
                                  "_hashCode": undefined,
                                  "altered": false,
                                  "root": ArrayMapNode {
                                    "entries": Array [
                                      Array [
                                        "a",
                                        "abc",
                                      ],
                                      Array [
                                        "c",
                                        Map {
                                          "_hashCode": undefined,
                                          "altered": false,
                                          "root": ArrayMapNode {
                                            "entries": Array [
                                              Array [
                                                "d",
                                                "dad",
                                              ],
                                              Array [
                                                "m",
                                                "mom",
                                              ],
                                              Array [
                                                "g",
                                                Map {
                                                  "_hashCode": undefined,
                                                  "altered": false,
                                                  "root": ArrayMapNode {
                                                    "entries": Array [
                                                      Array [
                                                        "x",
                                                        "new",
                                                      ],
                                                    ],
                                                  },
                                                  "size": 1,
                                                },
                                              ],
                                            ],
                                          },
                                          "size": 3,
                                        },
                                      ],
                                    ],
                                  },
                                  "size": 2,
                                },
                              ],
                            ],
                          },
                          "size": 3,
                        },
                      ],
                    ],
                  },
                  "size": 2,
                },
              ],
            ],
          },
          "size": 1,
        },
      ],
    ],
  },
  "size": 3,
}
`;
